{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from collections import Counter\n",
    "import pandas as pd\n",
    "import math\n",
    "import pysam\n",
    "import operator\n",
    "\n",
    "class PileupRecord:\n",
    "    \n",
    "    def __init__(self,line):\n",
    "        fields = line.split(\"\t\")\n",
    "        self.seq = fields[0]\n",
    "        self.pos = int(fields[1])\n",
    "        self.ref = fields[2]\n",
    "        self.rCount = int(fields[3])\n",
    "        self.rRes = fields[4]\n",
    "        self.qual = fields[5][:-1]\n",
    "        \n",
    "    def print(self):\n",
    "        print('{}    {}    {}    {}    {}    {}'.format(self.seq, self.pos, self.ref, self.rCount, self.rRes, self.qual))\n",
    "        \n",
    "\n",
    "\n",
    "def DetectPolymorphicSite(readResult,rCount):\n",
    "    \n",
    "    '''Detects variants at particular location, returns counts and types of them.\n",
    "       input: string, info about particular location alignment results from all reads\n",
    "       output: dataframe, columns: variant, count, type'''\n",
    "    \n",
    "    readResult = readResult.upper().replace(',','.')\n",
    "\n",
    "    irrelevant = list(set(re.findall(r'\\^[^\\.]', readResult)))\n",
    "    for s in irrelevant:\n",
    "        readResult = readResult.replace(s,'')\n",
    "    \n",
    "    occ = re.findall(r'[\\.][ACGT]*[+-][0-9][ACGT]*', readResult)\n",
    "    var = []\n",
    "    variants = list(set(occ))\n",
    "    varCounts = [occ.count(indel) for indel in variants]\n",
    "    varTypes = ['indel']*len(variants)\n",
    "    for i in range(0,len(variants)):\n",
    "        var.append([variants[i], varCounts[i], varTypes[i]])\n",
    "        \n",
    "    for s in variants:\n",
    "        readResult = readResult.replace(s,'')\n",
    "    \n",
    "    occ = re.findall(r'[AGCT]', readResult)\n",
    "    SNVs = list(set(occ))\n",
    "    SNVCounts = [occ.count(SNV) for SNV in SNVs]\n",
    "    SNVTypes = ['SNV']*len(SNVs)\n",
    "    for i in range(0,len(SNVs)):\n",
    "        var.append([SNVs[i], SNVCounts[i], SNVTypes[i]])\n",
    "\n",
    "    matchCount = len(re.findall(r'[\\.]', readResult))\n",
    "    var.append(['.', matchCount, 'match'])\n",
    "    \n",
    "    var.sort(key = lambda x: x[1], reverse = True)\n",
    "\n",
    "    if len(var) > 2:\n",
    "        if var[1][1] == 1 and rCount > 50:\n",
    "            return var[:1]\n",
    "        return var[:2]\n",
    "    else:\n",
    "        return var   \n",
    "\n",
    "\n",
    "def Genotyping(var,pro):\n",
    "    \n",
    "    '''Determines genotype.'''\n",
    "    \n",
    "    if len(var) == 1:\n",
    "        if var[0][2] == 'match':\n",
    "            genotype = (0,0)\n",
    "        else:\n",
    "            genotype = (1,1)\n",
    "        P = []\n",
    "    if len(var) == 2:\n",
    "        pr = [0]*3\n",
    "        k1 = var[0][1]\n",
    "        k2 = var[1][1]\n",
    "        # k1 = 40\n",
    "        # k2 = 1\n",
    "        p = pro\n",
    "        pr[0] = math.factorial(k1+k2)//math.factorial(k1)//math.factorial(k2)*(p**k1)*(1-p)**(k2) # a1a1\n",
    "        pr[1] = math.factorial(k1+k2)//math.factorial(k1+k2)//math.factorial(0)*(p**(k1+k2))*(1-p)**0 # a1a2\n",
    "        pr[2] = math.factorial(k1+k2)//math.factorial(k2)//math.factorial(k1)*p**k2*(1-p)**(k1) # a2a2\n",
    "    \n",
    "        P = [pr[0]/(pr[0]+pr[1]+pr[2]), pr[1]/(pr[0]+pr[1]+pr[2]), pr[2]/(pr[0]+pr[1]+pr[2])]\n",
    "        index, value = max(enumerate(P), key=operator.itemgetter(1))\n",
    "    \n",
    "        if var[0][2] == 'match': \n",
    "            if index == 0:\n",
    "                genotype = (0,0)\n",
    "            elif index == 1:\n",
    "                genotype = (0,1)\n",
    "            else:\n",
    "                genotype =(1,1)\n",
    "        elif var[1][2] == 'match':\n",
    "            if index == 0:\n",
    "                genotype = (1,1)\n",
    "            elif index == 1:\n",
    "                genotype = (0,1)\n",
    "            else:\n",
    "                genotype = (0,0)\n",
    "            P.reverse()\n",
    "        else:\n",
    "            if index == 0:\n",
    "                genotype = (1,1)\n",
    "            elif index == 1:\n",
    "                genotype = (1,2)\n",
    "            else:\n",
    "                genotype = (2,2)\n",
    "  \n",
    "    return genotype, P\n",
    "\n",
    "def DetermineAltsField(polymorphic_site):\n",
    "    \n",
    "    polymorphic_site = [elem for elem in polymorphic_site if(elem[0] != '.')]\n",
    "    if len(polymorphic_site) == 0:\n",
    "        alts = ['.']\n",
    "    else:\n",
    "        alts = [p[0] for p in polymorphic_site]\n",
    "    \n",
    "    return alts\n",
    "\n",
    "def AvgQual(qual):\n",
    "    qual_prob = []\n",
    "    for char in qual:\n",
    "        qual_prob.append(1 - 10**-((ord(char)-33)/10.0))\n",
    "    \n",
    "    return sum(qual_prob)/len(qual_prob)\n",
    "\n",
    "def Metrics(tp, fn, fp):\n",
    "    prec = tp/(tp+fp)\n",
    "    rec = tp/(tp+fn)\n",
    "    f1 = 2*prec*rec/(prec+rec)\n",
    "    \n",
    "    return prec, rec, f1\n",
    "\n",
    "pileup_file = open(\"merged-normal.pileup\")\n",
    "VCF_read = pysam.VariantFile(\"merged-normal.mpileup.called.vcf\", \"r\")\n",
    "\n",
    "\n",
    "    \n",
    "#counts = []\n",
    "#counts_true_pos = []\n",
    "#counts_false_pos = []\n",
    "#counts_false_neg = []\n",
    "#counts_true_neg = []\n",
    "\n",
    "probability = [0.6, 0.7, 0.8, 0.9, 0.95]\n",
    "#probability = [0.6,0.8]\n",
    "TP = []\n",
    "FN = []\n",
    "FP = []\n",
    "for pr in probability:\n",
    "    print(pr)\n",
    "    data_test = []\n",
    "    for rec in VCF_read.fetch():\n",
    "        data_test.append([rec.chrom, rec.pos])\n",
    "    \n",
    "    true_pos = 0\n",
    "    false_neg = 0\n",
    "    false_pos = 0\n",
    "    true_neg = 0\n",
    "\n",
    "    for line in pileup_file:\n",
    "        \n",
    "        rec = PileupRecord(line)\n",
    "        polymorphic_site = DetectPolymorphicSite(rec.rRes, rec.rCount)\n",
    "        genotype, probs = Genotyping(polymorphic_site, pr)\n",
    "        #counts.append([count[1] for count in polymorphic_site])\n",
    "        if (genotype != (0,0)) and ([rec.seq, rec.pos] in data_test):\n",
    "            true_pos += 1\n",
    "            data_test.remove([rec.seq, rec.pos])\n",
    "            #counts_true_pos.append([count[1] for count in polymorphic_site])\n",
    "        elif (genotype != (0,0)) and ([rec.seq, rec.pos] not in data_test): \n",
    "            false_pos += 1\n",
    "            #counts_false_pos.append([count[1] for count in polymorphic_site])\n",
    "        elif (genotype == (0,0)) and ([rec.seq, rec.pos] in data_test):\n",
    "            false_neg += 1\n",
    "            data_test.remove([rec.seq, rec.pos])\n",
    "            #counts_false_neg.append([count[1] for count in polymorphic_site])\n",
    "        elif (genotype == (0,0)) and ([rec.seq, rec.pos] not in data_test):\n",
    "            true_neg += 1\n",
    "            #counts_true_neg.append([count[1] for count in polymorphic_site])   \n",
    "\n",
    "    TP.append(true_pos)\n",
    "    FN.append(false_neg)\n",
    "    FP.append(false_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for tp, fn, fp in TP, FN, FP:\n",
    "    prec, re, f1 = Metrics(tp, fn, fp)\n",
    "    p.append(prec)\n",
    "    r.append(re)\n",
    "    f.append(f1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "fig=plt.figure(figsize =(10,6))\n",
    "plt.plot(probability, TP, label = 'true positives') \n",
    "plt.plot(probability, FP, label = 'false positives')\n",
    "plt.plot(probability, FN, label = 'false negatives')\n",
    "plt.xlabel('p') \n",
    "plt.ylabel('metrics') \n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "fig=plt.figure(figsize = (10,6))\n",
    "plt.plot(probability, p, label = 'precision') \n",
    "plt.plot(probability, r, label = 'recall')\n",
    "plt.plot(probability, f, label = 'f1 score')\n",
    "plt.xlabel('p') \n",
    "plt.ylabel('metrics') \n",
    "plt.legend()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
