{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "import pandas as pd\n",
    "import math\n",
    "import pysam\n",
    "import operator\n",
    "import re\n",
    "\n",
    "class PileupRecord:\n",
    "    \n",
    "    def __init__(self,line):\n",
    "        fields = line.split(\"\t\")\n",
    "        self.seq = fields[0]\n",
    "        self.pos = int(fields[1])\n",
    "        self.ref = fields[2]\n",
    "        self.rCount = int(fields[3])\n",
    "        self.rRes = fields[4]\n",
    "        self.qual = fields[5][:-1]\n",
    "        \n",
    "    def print(self):\n",
    "        print('{}    {}    {}    {}    {}    {}'.format(self.seq, self.pos, self.ref, self.rCount, self.rRes, self.qual))\n",
    "\n",
    "\n",
    "def VCFHeader():\n",
    "    VCFheader = pysam.VariantHeader()\n",
    "    VCFheader.add_sample('sample1')\n",
    "    VCFheader.add_line(\"##ALT=<ID=*,Description=Represents allele(s) other than observed.>\")\n",
    "    VCFheader.add_line(\"##FORMAT=<ID=GT,Number=1,Type=String,Description=Genotype>\")\n",
    "    VCFheader.add_line(\"##FORMAT=<ID=VAF,Number=1,Type=String,Description=Called genome probability>\")\n",
    "    \n",
    "def VCFBodyLine(rec, alts, genotype, p, VCF_out):\n",
    "        record = VCF_out.header.new_record()\n",
    "        record.contig = rec.seq\n",
    "        record.ref = rec.ref\n",
    "        record.alts = alts\n",
    "        record.pos = rec.pos\n",
    "        record.samples['HCC1143BL']['GT'] = genotype\n",
    "        record.samples['HCC1143BL']['VAF'] = str(p)\n",
    "    \n",
    "        return record\n",
    "    \n",
    "def DetermineIndelString(readResult):\n",
    "    \n",
    "    ind_num = [ind.start() for ind in (re.finditer(r'[0-9]', readResult))]\n",
    "    if len(ind_num) > 0:\n",
    "        string = readResult[2:ind_num[0]]\n",
    "        if len(ind_num) != 1:\n",
    "            for i in range(0,len(ind_num)-1):\n",
    "                string += int(readResult[ind_num[i]])*readResult[ind_num[i]+1:ind_num[i+1]]\n",
    "        string += int(readResult[ind_num[-1]])*readResult[ind_num[-1]+1:]\n",
    "    else:\n",
    "        string = readResult[2:]\n",
    "    return string\n",
    "\n",
    "def DetectPolymorphicSite(readResult):\n",
    "    \n",
    "    '''Detects variants at particular location, returns counts and types of them.\n",
    "       input: string, info about particular location alignment results from all reads\n",
    "       output: dataframe, columns: variant, count, type'''\n",
    "    \n",
    "    readResult = readResult.upper().replace(',','.')\n",
    "\n",
    "    irrelevant = list(set(re.findall(r'\\^[\\W][^\\.]', readResult)))\n",
    "    for s in irrelevant:\n",
    "        readResult = readResult.replace(s,'')\n",
    "    \n",
    "    occ = re.findall(r'[\\.][+-][ACGT]*[0-9]*[ACGT]*[0-9]*[ACGT]*', readResult)\n",
    "    var = []\n",
    "    variants = list(set(occ))\n",
    "    variants1 = [DetermineIndelString(variant) for variant in variants]\n",
    "    varCounts = [occ.count(indel) for indel in variants]\n",
    "    varTypes = ['indel']*len(variants1)\n",
    "    for i in range(0,len(variants1)):\n",
    "        var.append([variants1[i], varCounts[i], varTypes[i]])\n",
    "        \n",
    "    for s in variants:\n",
    "        readResult = readResult.replace(s,'')\n",
    "    \n",
    "    occ = re.findall(r'[AGCT]', readResult)\n",
    "    SNVs = list(set(occ))\n",
    "    SNVCounts = [occ.count(SNV) for SNV in SNVs]\n",
    "    SNVTypes = ['SNV']*len(SNVs)\n",
    "    for i in range(0,len(SNVs)):\n",
    "        var.append([SNVs[i], SNVCounts[i], SNVTypes[i]])\n",
    "\n",
    "    matchCount = len(re.findall(r'[\\.]', readResult))\n",
    "    var.append(['.', matchCount, 'match'])\n",
    "    \n",
    "    var.sort(key = lambda x: x[1], reverse = True)\n",
    "\n",
    "    if len(var) > 2:\n",
    "        return var[:2]\n",
    "    else:\n",
    "        return var \n",
    "\n",
    "\n",
    "def Genotyping(var,Qual):\n",
    "    \n",
    "    '''Determines genotype.'''\n",
    "    \n",
    "    if len(var) == 1:\n",
    "        if var[0][2] == 'match':\n",
    "            genotype = (0,0)\n",
    "        else:\n",
    "            genotype = (1,1)\n",
    "        P = []\n",
    "    if len(var) == 2:\n",
    "        pr = [0]*3\n",
    "        k1 = var[0][1]\n",
    "        k2 = var[1][1]\n",
    "        # k1 = 40\n",
    "        # k2 = 1\n",
    "        if var[0][2] == 'match' or var[0][2] == 'SNV':\n",
    "            p0 = 0.8\n",
    "        else:\n",
    "            p0 = 0.6\n",
    "            \n",
    "        if var[1][2] == 'match' or var[1][2] == 'SNV':\n",
    "            p1 = 0.8\n",
    "        else:\n",
    "            p1 = 0.6\n",
    "        p2 = (p0 + p1)/2\n",
    "\n",
    "        pr[0] = math.factorial(k1+k2)//math.factorial(k1)//math.factorial(k2)*(p0**k1)*(1-p0)**(k2) # a1a1\n",
    "        pr[1] = math.factorial(k1+k2)//math.factorial(k1+k2)//math.factorial(0)*(p2**(k1+k2))*(1-p2)**0 # a1a2\n",
    "        pr[2] = math.factorial(k1+k2)//math.factorial(k2)//math.factorial(k1)*p1**k2*(1-p1)**(k1) # a2a2\n",
    "    \n",
    "        P = [pr[0]/(pr[0]+pr[1]+pr[2]), pr[1]/(pr[0]+pr[1]+pr[2]), pr[2]/(pr[0]+pr[1]+pr[2])]\n",
    "        index, value = max(enumerate(P), key=operator.itemgetter(1))\n",
    "    \n",
    "        if var[0][2] == 'match': \n",
    "            if index == 0:\n",
    "                genotype = (0,0)\n",
    "            elif index == 1:\n",
    "                genotype = (0,1)\n",
    "            else:\n",
    "                genotype =(1,1)\n",
    "        elif var[1][2] == 'match':\n",
    "            if index == 0:\n",
    "                genotype = (1,1)\n",
    "            elif index == 1:\n",
    "                genotype = (0,1)\n",
    "            else:\n",
    "                genotype = (0,0)\n",
    "            P.reverse()\n",
    "        else:\n",
    "            if index == 0:\n",
    "                genotype = (1,1)\n",
    "            elif index == 1:\n",
    "                genotype = (1,2)\n",
    "            else:\n",
    "                genotype = (2,2)\n",
    "  \n",
    "    return genotype, P\n",
    "\n",
    "def DetermineAltsField(polymorphic_site):\n",
    "    \n",
    "    polymorphic_site = [elem for elem in polymorphic_site if(elem[0] != '.')]\n",
    "    if len(polymorphic_site) == 0:\n",
    "        alts = ['.']\n",
    "    else:\n",
    "        alts = [p[0] for p in polymorphic_site]\n",
    "    \n",
    "    return alts\n",
    "\n",
    "\n",
    "def AvgQual(qual):\n",
    "    qual_prob = []\n",
    "    for char in qual:\n",
    "        qual_prob.append(1 - 10**-((ord(char)-33)/10.0))\n",
    "    \n",
    "    return sum(qual_prob)/len(qual_prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "pileup_file = open(\"merged-normal.pileup\")\n",
    "VCF_read = pysam.VariantFile(\"merged-normal.mpileup.called.vcf\", \"r\")\n",
    "VCFheader = VCF_read.header\n",
    "VCFheader.add_line(\"##FORMAT=<ID=VAF,Number=1,Type=String,Description=Called genome probability>\")\n",
    "VCF_out = pysam.VariantFile(\"VCFout.vcf\", \"w\", header = VCFheader)\n",
    "\n",
    "for line in pileup_file:\n",
    "    rec = PileupRecord(line)\n",
    "    \n",
    "    polymorphic_site = DetectPolymorphicSite(rec.rRes,rec.rCount)\n",
    "    Qual = AvgQual(rec.qual)\n",
    "    genotype, probs = Genotyping(polymorphic_site, Qual)\n",
    "    \n",
    "    if genotype != (0,0):\n",
    "        alts = DetermineAltsField(polymorphic_site)\n",
    "        \n",
    "        VCF_line = VCFBodyLine(rec, alts, genotype, round(max(probs),4), VCF_out)\n",
    "        VCF_out.write(VCF_line)\n",
    "        \n",
    "VCF_out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
